plugins/embed-optimizer/class-embed-optimizer-tag-visitor.php

```diff
@@ -0,0 +1,111 @@
+<?php
+/**
+ * Tag visitor for Embed Optimizer.
+ *
+ * @package embed-optimizer
+ * @since n.e.x.t
+ */
+
+// Exit if accessed directly.
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+/**
+ * Tag visitor that optimizes embeds.
+ *
+ * @since 0.1.0
+ * @access private
+ */
+final class Embed_Optimizer_Tag_Visitor {
+
+	/**
+	 * URL Metrics Group Collection.
+	 *
+	 * @var OD_URL_Metrics_Group_Collection
+	 */
+	protected $url_metrics_group_collection;
+
+	/**
+	 * Link Collection.
+	 *
+	 * @var OD_Link_Collection
+	 */
+	protected $link_collection;
+
+	/**
+	 * Whether the lazy-loading script was added to the body.
+	 *
+	 * @var bool
+	 */
+	protected $added_lazy_script = false;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param OD_URL_Metrics_Group_Collection $url_metrics_group_collection URL Metrics Group Collection.
+	 * @param OD_Link_Collection              $link_collection              Link Collection.
+	 */
+	public function __construct( OD_URL_Metrics_Group_Collection $url_metrics_group_collection, OD_Link_Collection $link_collection ) {
+		$this->url_metrics_group_collection = $url_metrics_group_collection;
+		$this->link_collection              = $link_collection;
+	}
+
+	/**
+	 * Visits a tag.
+	 *
+	 * @param OD_HTML_Tag_Processor $processor Processor.
+	 * @return bool Whether the visit or visited the tag.
+	 */
+	public function __invoke( OD_HTML_Tag_Processor $processor ): bool {
+		if ( ! (
+			'FIGURE' === $processor->get_tag()
+			&&
+			$processor->has_class( 'wp-block-embed' )
+		) ) {
+			return false;
+		}
+
+		$original_bookmarks = $processor->get_bookmark_names();
+
+		$max_intersection_ratio = $this->url_metrics_group_collection->get_element_max_intersection_ratio( $processor->get_xpath() );
+
+		if ( $max_intersection_ratio > 0 ) {
+
+			$preconnect_hrefs = array();
+			// TODO: Add more cases.
+			if ( $processor->has_class( 'wp-block-embed-youtube' ) ) {
+				$preconnect_hrefs[] = 'https://www.youtube.com';
+				$preconnect_hrefs[] = 'https://i.ytimg.com';
+			} elseif ( $processor->has_class( 'wp-block-embed-twitter' ) ) {
+				$preconnect_hrefs[] = 'https://syndication.twitter.com';
+				$preconnect_hrefs[] = 'https://pbs.twimg.com';
+			} elseif ( $processor->has_class( 'wp-block-embed-wordpress-tv' ) ) {
+				$preconnect_hrefs[] = 'https://video.wordpress.com';
+				$preconnect_hrefs[] = 'https://public-api.wordpress.com';
+				$preconnect_hrefs[] = 'https://videos.files.wordpress.com';
+			}
```

Comment:
@adamsilverstein Any other embeds you'd recommend for the initial merge of this?

Reply:
To be updated in subsequent PR based on the Colab data on WP embeds.

Reply:
The most common 10 embed types used on WordPress sites and the percent of sites with embeds hat use that type:

![image](https://github.com/WordPress/performance/assets/2676022/08bc7e26-1835-4af6-8fdc-e2f8a6ace673)

data from httparchive, queries in [colab](https://colab.sandbox.google.com/drive/1nSpg3qoCLY-cBTV2zOUkgUCU7R7X2f_R?resourcekey=0-MgT7Ur0pT__vw-5_AHjgWQ#scrollTo=utZv59sXzXvS)

Reply:
Awesome! What is `embed-handler` and `amazon`?

Reply:
`amazon` might refer to Amazon Kindle embeds.

`embed-handler` sounds odd though 🤔 Maybe coming from custom embed handlers added via `wp_embed_register_handler` where the developer forgot to change the class name?

Reply:
yes, we support amazon embeds! `embed` is when you embed one WordPress post in another post (example - https://oembeds-test.instawp.xyz/test-wordpress-post/)

Reply:
Full [list of oembeds](https://github.com/WordPress/wordpress-develop/blob/9cff1de6b3adcf8767bfc60c4e71b5cce81f0e1b/src/wp-includes/class-wp-oembed.php#L53) in core.

Reply:
Excellent. I've fleshed out the list of embeds to preconnect in https://github.com/WordPress/performance/pull/1302/commits/6c3a9b431b368e51a1bba04da5d11e3741ca6a53. I went down to Soundcloud.